typegen/src/tests/mod.rs

```diff
@@ -0,0 +1,1143 @@
+                    }
+                }
+            }
+        }
+    };
+
+    assert_eq!(code.to_string(), expected_code.to_string());
+}
+
+#[test]
+fn dont_force_struct_names_camel_case() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct AB;
+
+    let code = Testgen::new().with::<AB>().gen_tests_mod(subxt_settings());
+
+    let expected_code = quote! {
+        pub mod tests {
+            use super::root;
+
+            #[derive(::subxt_path::ext::codec::Decode, ::subxt_path::ext::codec::Encode, ::subxt_path::ext::scale_decode::DecodeAsType, ::subxt_path::ext::scale_encode::EncodeAsType, Debug)]
+            #[codec(crate = ::subxt_path::ext::codec)]
+            #[decode_as_type(crate_path = ":: subxt_path :: ext :: scale_decode")]
+            #[encode_as_type(crate_path = ":: subxt_path :: ext :: scale_encode")]
+            pub struct AB;
+        }
+    };
+
+    assert_eq!(code.to_string(), expected_code.to_string());
+}
+
+#[test]
+fn apply_user_defined_derives_for_all_types() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct A(B);
+
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct B;
+
+    let mut settings = subxt_settings();
+    settings.derives.extend_for_all(
+        [parse_quote!(Clone), parse_quote!(Eq)],
+        [parse_quote!(#[some_attribute])],
+    );
+
+    let code = Testgen::new().with::<A>().gen_tests_mod(settings);
+
+    let expected_code = quote! {
+        pub mod tests {
+            use super::root;
+
+            #[derive(::subxt_path::ext::codec::Decode, ::subxt_path::ext::codec::Encode, ::subxt_path::ext::scale_decode::DecodeAsType, ::subxt_path::ext::scale_encode::EncodeAsType, Clone, Debug, Eq)]
+            #[codec(crate = ::subxt_path::ext::codec)]
+            #[decode_as_type(crate_path = ":: subxt_path :: ext :: scale_decode")]
+            #[encode_as_type(crate_path = ":: subxt_path :: ext :: scale_encode")]
+            #[some_attribute]
+            pub struct A(pub root :: scale_typegen :: tests :: B,);
+
+            #[derive(::subxt_path::ext::codec::Decode, ::subxt_path::ext::codec::Encode, ::subxt_path::ext::scale_decode::DecodeAsType, ::subxt_path::ext::scale_encode::EncodeAsType, Clone, Debug, Eq)]
+            #[codec(crate = ::subxt_path::ext::codec)]
+            #[decode_as_type(crate_path = ":: subxt_path :: ext :: scale_decode")]
+            #[encode_as_type(crate_path = ":: subxt_path :: ext :: scale_encode")]
+            #[some_attribute]
+            pub struct B;
+        }
+    };
+
+    assert_eq!(code.to_string(), expected_code.to_string());
+}
+
+#[test]
+fn apply_user_defined_derives_for_specific_types() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct A(B);
+
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct B(C);
+
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct C;
+
+    let mut settings = subxt_settings();
+    settings.derives.extend_for_all([parse_quote!(Eq)], []);
```

Comment:
Would it be more specific to:
- Have structures for input of the `settings.derives` to have named fields to easily identify which are derives and attributes
- Rename `extend_for_type` to `extend` since we have introduced named fields?

```rust
derives.extend_for_all(
    ExtendAll {
        derives: [...],
        attributes: [],
    });

derives.extend(
    Extend {
        ty: "scale_type::tests::B",
        derives: vec!["Hash"],
        attribute: vec!["some_attr"],
    }
)
```

Another alternative might be
```rust
derives.extend(ExtendDerive::for_all().derives(..).attributes(..));
derives.extend(ExtendDerive::ty(..).derives(..).attributes(..));
```

Reply:
or just separate the derives and attribute adding so it's more like

```rust
derives.add_attribute_for_all(attr);
derives.add_attribute_for(path, attr);

derives.add_derive_for_all(derive);
derives.add_derive_for(path, derive);
```